File: hackerrank/challenges/FrequencyQueries.java
package challenges;

import java.util.HashMap;
import java.util.LinkedList;
import java.util.List;

public class FrequencyQueries {
    public static void main(String[] args) {
    }

    // Complete the freqQuery function below.
    static List<Integer> freqQuery(List<List<Integer>> queries) {
        List<Integer> result = new LinkedList<>();
//        HashMap<Integer, Integer> map = new HashMap<>();
//        HashMap<Integer, Integer> freq = new HashMap<>();
//
//        for (List<Integer> query : queries) {
//            Integer command = query.get(0);
//            Integer value = query.get(1);
//            Integer prev = map.getOrDefault(value, 0);
//            Integer next = 0;
//
//            switch (command)
//            {
//                case 1:
//                    freq.put(prev, freq.getOrDefault(prev, 0) - 1);
//                    next = prev + 1;
//                    freq.put(next, freq.getOrDefault(prev, 0) + 1);
//                    map.put(value, next);
//                    break;
//                case 2:
//                    freq.put(prev, freq.getOrDefault(prev, 0) - 1);
//                    next = prev - 1;
//                    freq.put(next, freq.getOrDefault(prev, 0) + 1);
//                    map.put(value, next);
//                    break;
//                case 3:
//                    if (freq.getOrDefault(value, 0) > 0) {
//                        result.add(1);
//                    } else {
//                        result.add(0);
//                    }
//                    break;
//            }
//        }
//
        return result;
    }
}

File: hackerrank/challenges/FrequencyQueries.js
const INSERT_COMMAND = 1;
const DELETE_COMMAND = 2;
const CHECK_COMMAND = 3;

// Complete the freqQuery function below.
function freqQuery(queries) {
    let result = [];
    let hash = {};
    let freq = [];

    for (let i = 0; i < queries.length; i += 1) {
        let [command, value] = queries[i];
        let prev = hash[value] || 0;

        switch (command) {
            case INSERT_COMMAND:
                hash[value] = prev + 1;
                freq[prev] = (freq[prev] || 0) - 1;
                freq[prev + 1] = (freq[prev + 1] || 0) + 1;
                break;
            case DELETE_COMMAND:
                if (prev <= 0) break;

                hash[value] = prev - 1;
                freq[prev - 1] += 1;
                freq[prev] -= 1;
                break;
            case CHECK_COMMAND:
                result.push(freq[value] > 0 ? 1 : 0)
                break;
        }
    }

    return result;
}

console.log(freqQuery([[1, 5], [1, 6], [3, 2], [1, 10], [1, 10], [1, 6], [2, 5], [3, 2]])) // 0, 1
